packages/pwa-kit-dev/src/configs/webpack/overrides-plugin.js

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2023, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import path from 'path'
+import glob from 'glob'
+
+/**
+ * @class OverridesResolverPlugin
+ *
+ *  This plugin provides the "Overrides" behavior of the Template Extensibility feature,
+ *  allowing third party implementations that depend on an npm module for the base implementation
+ *  and then overriding only specific files
+ */
+class OverridesResolverPlugin {
+    /**
+     *
+     * @param options
+     * @param {string} options.overridesDir path to application base
+     * @param {string[]} options.overrides paths to overrides
+     * @param {string} options.projectDir path to project directory
+     */
+    constructor(options) {
+        this.overridesDir = options.overridesDir || ''
+        this.extends = options.extends || []
+        this.projectDir = options.projectDir
+        this._allSearchDirs = [this.projectDir + this.overridesDir, ...this.extends]
+        this.pkg = require(path.resolve(this.projectDir, 'package.json'))
+        this.extendsHashMap = new Map()
+
+        const OVERRIDES_EXTENSIONS = '.+(js|jsx|ts|tsx|svg|jpg|jpeg)'
```

Comment:
file extensions that can override underlying template (e.g. `routes.tsx` can override `routes.jsx`)

Reply:
Was the decision made to do this because we the argument might be made that you want to write your project in typescript but extend a non-typescript project?

Are there any scenarios where this decision might not work? E.g. to files with the same name in the extendable project but different file terminations?

Will there be a precedence to override the file with the making file termination then move on to one that might be supported after?

Reply:
I don't understand your question here

> Are there any scenarios where this decision might not work? E.g. to files with the same name in the extendable project but different file terminations?

Are you pointing out that you can do `routes.png` as a replacement for `routes.jsx` and footgun yourself as an implementor? Yes. We could make this more sophisticated where we have "groups" of file extensions that allow other files as overrides. E.g. `const TEMPLATE_OVERRIDES = ['jsx', 'tsx']` and `const JS_OVERRIDES = ['js', 'ts']` to narrow the scope, but I also think we could do this in a follow up

At the end of the day though, I think an allow-list is required here, but open to suggestions if you have ideas for doing this without letting customers foot-gun themselves more than they already are capable of

Reply:
if the overrides have a mismatch of the file extension, we can log warnings